### PR TITLE
✨ Save delays for single stops and refresh it automatically

### PR DESCRIPTION
**Pull request by MrKrisKrisu**
image

### CODE REVIEW

File: database/migrations/2021_01_26_203951_create_train_stop_overs_table.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+use App\Models\TrainStopOver;
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+class CreateTrainStopOversTable extends Migration
+{
+    public function up(): void {
+        Schema::create('train_stop_overs', function(Blueprint $table) {
```
Comment (HerrLevin) [inline database/migrations/2021_01_26_203951_create_train_stop_overs_table.php]: train_stopovers

File: database/migrations/2021_01_26_203951_create_train_stop_overs_table.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+use App\Http\Controllers\HafasController;
+use App\Models\TrainStopOver;
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+class CreateTrainStopOversTable extends Migration
```
Comment (HerrLevin) [inline database/migrations/2021_01_26_203951_create_train_stop_overs_table.php]: CreateTrainStopoversTable

File: app/Models/HafasTrip.php [hunk 3]
```diff
@@ -28,4 +29,9 @@ public function originStation() {
     public function destinationStation() {
         return $this->belongsTo(TrainStation::class, 'destination', 'ibnr');
     }
+
+    public function stopoversNEW(): HasMany {
+        //TODO: Rename to ->stopovers when old attribute is away
+        return $this->hasMany(TrainStopover::class, 'trip_id', 'trip_id')->orderBy('arrival_planned');
+    }
 }
```
Comment (HerrLevin) [inline app/Models/HafasTrip.php]: !!!

File: app/Http/Controllers/HafasController.php [hunk 4]
```diff
@@ -171,25 +172,67 @@ private static function fetchHafasTrip(string $tripID, string $lineName): HafasT
...
+            //us delay-data if the train already passed this station
+            //so.. just save data we really got. :)
+            $updatePayload = [];
+
+            if ($stopover->plannedArrival != null) {
+                $updatePayload['arrival_planned'] = $stopover->plannedArrival;
+            }
+            if ($stopover->arrival != null && Carbon::parse($stopover->arrival)->isFuture()) {
+                $updatePayload['arrival_real'] = $stopover->arrival;
+                if ($stopover->plannedArrivalPlatform != null) {
```
Comment (HerrLevin) [inline app/Http/Controllers/HafasController.php]: every nested `if` can be removed

File: app/Http/Controllers/HafasController.php [hunk 5]
```diff
@@ -171,25 +172,67 @@ private static function fetchHafasTrip(string $tripID, string $lineName): HafasT
...
+                }
+                if ($stopover->arrivalPlatform != null) {
+                    $updatePayload['arrival_platform_real'] = $stopover->arrivalPlatform;
+                }
+            }
+            if ($stopover->plannedDeparture != null) {
+                $updatePayload['departure_planned'] = $stopover->plannedDeparture;
+            }
+            if ($stopover->departure != null && Carbon::parse($stopover->departure)->isFuture()) {
+                $updatePayload['departure_real'] = $stopover->departure;
```
Comment (HerrLevin) [inline app/Http/Controllers/HafasController.php:223]: every nested if can be removed

File: app/Http/Controllers/HafasController.php [hunk 6]
```diff
@@ -171,25 +172,67 @@ private static function fetchHafasTrip(string $tripID, string $lineName): HafasT
...
+
+        foreach ($tripJson->stopovers as $stopover) {
+            $hafasStop = self::parseHafasStopObject($stopover->stop);
+
+            //This array is a workaround because Hafas doesn't give
+            //us delay-data if the train already passed this station
+            //so.. just save data we really got. :)
+            $updatePayload = [];
+
+            if ($stopover->plannedArrival != null) {
```
Comment (HerrLevin) [inline app/Http/Controllers/HafasController.php]: You can remove every `null`-check and directly set the variables if you remove every null-value with `array_filter($arr, static function($var){return $var !== null;} );` or sth. else

File: app/Http/Controllers/FrontendStatusController.php [hunk 7]
```diff
@@ -226,7 +226,20 @@ public function usageboard(Request $request) {
         ]);
     }
 
+    /**
+     * @param $status
+     * @return mixed
+     * @todo move to Status Model and return StopOver instead of String
+     */
     public static function nextStation(&$status) {
```
Comment (HerrLevin) [inline app/Http/Controllers/FrontendStatusController.php:238]: Can we just move this to JavaScript already? 😫

**Review by HerrLevin — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/FrontendStatusController.php (hunk 7 above)
Comment (MrKrisKrisu) [reply app/Http/Controllers/FrontendStatusController.php:238]: Do it if you want. I'm out of that.

**Review by HerrLevin — APPROVED** [review]

**Pull request merged**